# Введение в обработку текста на естественном языке

Материалы:
* Макрушин С.В. Лекция 9: Введение в обработку текста на естественном языке\
* https://realpython.com/nltk-nlp-python/
* https://scikit-learn.org/stable/modules/feature_extraction.html

## Задачи для совместного разбора

In [5]:
!pip install pymorphy2

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.5/55.5 kB 2.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.2/8.2 MB 48.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for docopt: filename=docopt-0.6.2-py2.py3-none-any.whl size=13707 sha256=23cfb74f714054405e79638e28fa1bf73a683a8d30ff7765e7d5649626e7e4f4
  Stored in directory: /root/.cache/pip/wheels/fc/ab/d4/5da2067ac95b36618c629a5f93f809425700506f72c9732fac
Successfully built docopt


In [6]:
from sklearn.feature_extraction.text import CountVectorizer
import pymorphy2
from nltk.metrics.distance import edit_distance

In [13]:
!pip install nltk
import nltk
nltk.download('punkt')

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

1. Считайте слова из файла `litw-win.txt` и запишите их в список `words`. В заданном предложении исправьте все опечатки, заменив слова с опечатками на ближайшие (в смысле расстояния Левенштейна) к ним слова из списка `words`. Считайте, что в слове есть опечатка, если данное слово не содержится в списке `words`. 

In [ ]:
text = '''с велечайшим усилием выбравшись из потока убегающих людей Кутузов со свитой уменьшевшейся вдвое поехал на звуки выстрелов русских орудий'''

In [14]:
with open('litw-win.txt', 'r', encoding='windows-1251') as file:
    words = file.read().split()

sentence ="с велечайшим усилием выбравшись из потока убегающих людей Кутузов со свитой уменьшевшейся вдвое поехал на звуки выстрелов русских орудий"

def correct_typos(sentence, words):
    corrected_sentence = []
    for word in nltk.word_tokenize(sentence):
        if word not in words:
            closest_word = min(words, key=lambda x: edit_distance(word, x))
            corrected_sentence.append(closest_word)
        else:
            corrected_sentence.append(word)
    return ' '.join(corrected_sentence)

corrected_sentence = correct_typos(sentence, words)
print(corrected_sentence)

с величайшим усилием выбравшись из потока убегающих людей кутузов со свитой уменьшившейся вдвое поехал на звуки выстрелов русских орудий


2. Разбейте текст из формулировки задания 1 на слова; проведите стемминг и лемматизацию слов.

In [24]:
from nltk.stem import SnowballStemmer, WordNetLemmatizer
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [25]:
text = "с велечайшим усилием выбравшись из потока убегающих людей Кутузов со свитой уменьшевшейся вдвое поехал на звуки выстрелов русских орудий"
words = text.split()
stemmer = SnowballStemmer('russian')
stemmed_words = [stemmer.stem(word) for word in words]
lemmatizer = WordNetLemmatizer()
lemmatized_words = [lemmatizer.lemmatize(word, pos='v') for word in words]

print("Слова:", words)
print("Стемминг:", stemmed_words)
print("Лемматизация:", lemmatized_words)

Слова: ['с', 'велечайшим', 'усилием', 'выбравшись', 'из', 'потока', 'убегающих', 'людей', 'Кутузов', 'со', 'свитой', 'уменьшевшейся', 'вдвое', 'поехал', 'на', 'звуки', 'выстрелов', 'русских', 'орудий']
Стемминг: ['с', 'велечайш', 'усил', 'выбра', 'из', 'поток', 'убега', 'люд', 'кутуз', 'со', 'свит', 'уменьшевш', 'вдво', 'поеха', 'на', 'звук', 'выстрел', 'русск', 'оруд']
Лемматизация: ['с', 'велечайшим', 'усилием', 'выбравшись', 'из', 'потока', 'убегающих', 'людей', 'Кутузов', 'со', 'свитой', 'уменьшевшейся', 'вдвое', 'поехал', 'на', 'звуки', 'выстрелов', 'русских', 'орудий']


3. Преобразуйте предложения из формулировки задания 1 в векторы при помощи `CountVectorizer`.

In [17]:
from sklearn.feature_extraction.text import CountVectorizer

In [26]:
text = "с велечайшим усилием выбравшись из потока убегающих людей Кутузов со свитой уменьшевшейся вдвое поехал на звуки выстрелов русских орудий"

words = text.split()
new_text = " ".join(words)
vectorizer = CountVectorizer()
vectorized_text = vectorizer.fit_transform([new_text])

print(vectorized_text.toarray())

[[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]]


## Лабораторная работа 9

### Расстояние редактирования

1.1 Загрузите предобработанные описания рецептов из файла `preprocessed_descriptions.csv`. Получите набор уникальных слов `words`, содержащихся в текстах описаний рецептов (воспользуйтесь `word_tokenize` из `nltk`). 

In [30]:
import pandas as pd
from nltk.tokenize import word_tokenize

In [ ]:
df = pd.read_csv('preprocessed_descriptions.csv')
unique_words = set()
words_list = []
descriptions = [word_tokenize(d) for d in df["preprocessed_descriptions"].tolist() if isinstance(d, str)]

[[unique_words.add(word) for word in desc] for desc in descriptions]
[[words_list.append(word) for word in desc] for desc in descriptions]

for desc in descriptions:
  print(desc)

1.2 Сгенерируйте 5 пар случайно выбранных слов и посчитайте между ними расстояние редактирования.

In [35]:
import random

In [38]:
unique_words_set = set()
unique_words_list = list()

words = [word_tokenize(item) for item in df["preprocessed_descriptions"].to_list() if isinstance(item, str)]
[[unique_words_set.add(x) for x in item] for item in words]
[[unique_words_list.append(x) for x in item] for item in words]
data = random.sample(list(unique_words_set), 10)

for i in range(0, len(data), 2):
    word1, word2 = data[i], data[i+1]
    print(edit_distance(word1, word2), word1, word2)

45 loudly httpwwwcookingwithtraderjoescomsamplerecipe1html
9 stayovers switzerland
6 wowwee egypt
12 breakfastits 309199yum
10 subscribed lacalamita


1.3 Напишите функцию, которая для заданного слова `word` возвращает `k` ближайших к нему слов из списка `words` (близость слов измеряется с помощью расстояния Левенштейна)

In [40]:
from typing import Set, List

def get_similar_words(word: str, k: int, words: Set[str]) -> List[str]:
    """Функция для возврата k подобных слов для word из коллекции words_data"""
    buf_tuple = [(edit_distance(word, item), item) for item in words]
    buf_tuple.sort(key=lambda x: x[0])
    return [x[1] for x in buf_tuple[:k]]

get_similar_words("seedless", 11, unique_words_set)

['seedless',
 'needless',
 'needles',
 'endless',
 'restless',
 'needle',
 'settles',
 'skinless',
 'peelers',
 'seeded',
 'meatless']

### Стемминг, лемматизация

2.1 На основе результатов 1.1 создайте `pd.DataFrame` со столбцами: 
    * word
    * stemmed_word 
    * normalized_word 

Столбец `word` укажите в качестве индекса. 

Для стемминга воспользуйтесь `SnowballStemmer`, для нормализации слов - `WordNetLemmatizer`. Сравните результаты стемминга и лемматизации.

In [42]:
words_dataframe = pd.DataFrame(unique_words_set)
words_dataframe.columns = ['word']
words_dataframe['stemmed_word'] = words_dataframe.apply(lambda x: stemmer.stem(x["word"]), axis=1)
words_dataframe['normalized_word'] = words_dataframe.apply(lambda x: lemmatizer.lemmatize(x["word"], "v"), axis=1)

filtered_words = words_dataframe[(words_dataframe["word"] != words_dataframe["normalized_word"]) & (words_dataframe["stemmed_word"] != words_dataframe["normalized_word"])]
filtered_words

word stemmed_word normalized_word
7         bought       bought             buy
18       amazing      amazing           amaze
48     requested    requested         request
72       directs      directs          direct
78         posts        posts            post
...          ...          ...             ...
32834  imitating    imitating         imitate
32835   shucking     shucking           shuck
32836     laying       laying             lay
32857     asking       asking             ask
32862     pearls       pearls           pearl

[3899 rows x 3 columns]

2.2. Удалите стоп-слова из описаний рецептов. Какую долю об общего количества слов составляли стоп-слова? Сравните топ-10 самых часто употребляемых слов до и после удаления стоп-слов.

In [46]:
from nltk.corpus import stopwords
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [52]:
words_preprocessed = [word_tokenize(item) for item in df["preprocessed_descriptions"].to_list() if isinstance(item, str)]
words_list = [word for sublist in words_preprocessed for word in sublist]
stopwords_set = set(stopwords.words('english'))

words_filtered = [word for word in words_list if word not in stopwords_set]
diff = round(len(words_filtered)/len(words_list)*100,2)
print(f"Слов: {len(words_list)}\nПосле удаления: {len(words_filtered)}\nПроцент стоп-слов: {diff}%"+"\n")

freq = nltk.FreqDist(words_list)
print("10 слов (с стоп-словами):")
for word, number in freq.most_common(10): print(f"{number} -> {word}")
print()
freq_filtered = nltk.FreqDist(words_filtered)
print("10 слов (без стоп-слов):")
for word, number in freq_filtered.most_common(10): print(f"{number} -> {word}")


Слов: 1069254
После удаления: 580889
Процент стоп-слов: 54.33%

10 слов (с стоп-словами):
40072 -> the
34951 -> a
30245 -> and
26859 -> this
24836 -> i
23471 -> to
20285 -> is
19756 -> it
18364 -> of
15939 -> for

10 слов (без стоп-слов):
14871 -> recipe
6326 -> make
5137 -> time
4620 -> use
4430 -> great
4167 -> like
4152 -> easy
3872 -> one
3810 -> made
3791 -> good


### Векторное представление текста

3.1 Выберите случайным образом 5 рецептов из набора данных. Представьте описание каждого рецепта в виде числового вектора при помощи `TfidfVectorizer`

In [53]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [64]:
data = df.sample(5)
vectorizer = TfidfVectorizer()
for i, row in data.iterrows():
    description = data["preprocessed_descriptions"]
    vector = vectorizer.fit_transform(data["preprocessed_descriptions"])
    print(f"Рецепт {i+1}:\n", vector.toarray(), "\n")

Рецепт 643:
 [[0.         0.13588446 0.         0.13588446 0.2296648  0.
  0.13588446 0.         0.         0.         0.21926145 0.10963073
  0.         0.13588446 0.         0.         0.         0.
  0.         0.         0.         0.13588446 0.         0.
  0.         0.         0.07655493 0.         0.09100341 0.
  0.10963073 0.         0.         0.         0.         0.
  0.13588446 0.13588446 0.         0.13588446 0.         0.
  0.         0.         0.         0.         0.13588446 0.
  0.         0.         0.         0.         0.         0.
  0.         0.10963073 0.13588446 0.         0.18200681 0.
  0.         0.         0.         0.         0.         0.13588446
  0.         0.         0.         0.         0.         0.13588446
  0.         0.         0.         0.         0.         0.
  0.         0.         0.27176892 0.         0.         0.
  0.         0.18200681 0.         0.         0.13588446 0.
  0.27176892 0.         0.         0.13588446 0.         0.
  0

3.2 Вычислите близость между каждой парой рецептов, выбранных в задании 3.1, используя косинусное расстояние (`scipy.spatial.distance.cosine`) Результаты оформите в виде таблицы `pd.DataFrame`. В качестве названий строк и столбцов используйте названия рецептов.

In [65]:
from scipy.spatial.distance import cosine

In [73]:
data

Unnamed: 0                                               name  \
642           642                     amazing white trash puff balls   
23360       23360  sarasota s savory roasted chicken  apples   on...   
26817       26817                                       swiss   zopf   
1032         1032                        applesauce streusel muffins   
7423         7423            colorado style green chili  slow cooker   

                               preprocessed_descriptions  
642    these are wonderful my cousin brought them as ...  
23360  easy flavorful roasted chicken with very littl...  
26817  this recipe is by victoria marler and has been...  
1032   what a great way to use fresh apples  the smel...  
7423   this is a recipe i came up by using several di...

In [84]:
tfidf = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf.fit_transform(data["preprocessed_descriptions"])

similarity_matrix = []
for i in range(len(data)):
    row = []
    for j in range(len(data)):
        if i == j:
            row.append(1.0)
        else:
            cosine_distance = cosine(tfidf_matrix[i].toarray()[0], tfidf_matrix[j].toarray()[0])
            similarity = 1 - cosine_distance
            row.append(similarity)
    similarity_matrix.append(row)
similarity_df = pd.DataFrame(similarity_matrix, columns=data['name'], index=data['name'])
similarity_df

name                                                amazing white trash puff balls  \
name                                                                                 
amazing white trash puff balls                                            1.000000   
sarasota s savory roasted chicken  apples   onions                        0.055044   
swiss   zopf                                                              0.129166   
applesauce streusel muffins                                               0.000000   
colorado style green chili  slow cooker                                   0.072054   

name                                                sarasota s savory roasted chicken  apples   onions  \
name                                                                                                     
amazing white trash puff balls                                                               0.055044    
sarasota s savory roasted chicken  apples   onions                                           1.000000    
swiss   zopf                                                                                 0.044458    
applesauce streusel muffins                                                                  0.074981    
colorado style green chili  slow cooker                                                      0.021677    

name                                                swiss   zopf  \
name                                                               
amazing white trash puff balls                          0.129166   
sarasota s savory roasted chicken  apples   onions      0.044458   
swiss   zopf                                            1.000000   
applesauce streusel muffins                             0.000000   
colorado style green chili  slow cooker                 0.033879   

name                                                applesauce streusel muffins  \
name                                                                              
amazing white trash puff balls                                         0.000000   
sarasota s savory roasted chicken  apples   onions                     0.074981   
swiss   zopf                                                           0.000000   
applesauce streusel muffins                                            1.000000   
colorado style green chili  slow cooker                                0.031882   

name                                                colorado style green chili  slow cooker  
name                                                                                         
amazing white trash puff balls                                                     0.072054  
sarasota s savory roasted chicken  apples   onions                                 0.021677  
swiss   zopf                                                                       0.033879  
applesauce streusel muffins                                                        0.031882  
colorado style green chili  slow cooker                                            1.000000

3.3 Какие рецепты являются наиболее похожими? Прокомментируйте результат (словами).

Результаты показывают, что наиболее похожими рецептами являются "swiss zopf" и "amazing white trash puff balls". Они имеют наивысшую близость косинусного расстояния в 0.129166.